core/view_index.h

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Neil Vaytet
+#ifndef VIEW_INDEX_H
+#define VIEW_INDEX_H
+
+#include "dimensions.h"
+
+namespace scipp::core {
+
+class ViewIndex {
+public:
+  ViewIndex(const Dimensions &parentDimensions,
+            const Dimensions &subdimensions) {
+    m_dims = parentDimensions.shape().size();
+    for (scipp::index d = 0; d < m_dims; ++d)
+      m_extent[d] = parentDimensions.size(m_dims - 1 - d);
+    scipp::index factor{1};
+    for (scipp::index i = subdimensions.shape().size() - 1; i >= 0; --i) {
+      const auto dimension = subdimensions.label(i);
+      if (parentDimensions.contains(dimension)) {
+        m_offsets[m_subdims] = m_dims - 1 - parentDimensions.index(dimension);
+        m_factors[m_subdims] = factor;
+        ++m_subdims;
+      }
+      factor *= subdimensions.size(i);
+    }
+    scipp::index offset{1};
+    for (scipp::index d = 0; d < m_dims; ++d) {
+      setIndex(offset);
+      m_delta[d] = m_index;
+      if (d > 0) {
+        setIndex(offset - 1);
+        m_delta[d] -= m_index;
+        for (scipp::index d2 = 0; d2 < d; ++d2)
+          m_delta[d] -= m_delta[d2];
+      }
+      offset *= m_extent[d];
+    }
+    setIndex(0);
+  }
+
+  void increment() {
+    m_index += m_delta[0];
+    ++m_coord[0];
+    scipp::index d = 0;
+    while ((m_coord[d] == m_extent[d]) && (d < NDIM_MAX - 1)) {
```

Comment:
I'd be interested to see your benchmarks. In the original multi-index implementation it made a big difference how this was implemented. Having `increment` as a large function prevented inlining, making it really slow (I split it into 2 functions). Is `increment` getting inlined with your implementation?